bin/commands/indexDump.js

```diff
@@ -0,0 +1,117 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2018 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable no-console */
+
+const
+  fs = require('fs'),
+  ColorOutput = require('./colorOutput'),
+  path = require('path'),
+  ndjson = require('ndjson'),
+  getSdk = require('./getSdk');
+
+//@TODO use 'recursive: true' with node.js 10
+function mkdirp (fullPath) {
+  const parts = fullPath.split(path.sep);
+
+  if (parts.length > 1) {
+    mkdirp(parts.slice(0, parts.length - 1).join('/'));
+  }
+
+  if (! fs.existsSync(fullPath)) {
+    fs.mkdirSync(fullPath);
+  }
+}
+
+function writeData(stream, data) {
+  return new Promise(resolve => stream.write(data, () => resolve()));
+}
+
+async function dumpCollection (sdk, index, collection, batchSize, directoryPath) {
+  const
+    filename = `${directoryPath}/${index}--${collection}--data.jsonl`,
+    writeStream = fs.createWriteStream(filename),
+    waitWrite = new Promise(resolve => writeStream.on('finish', resolve)),
+    ndjsonStream = ndjson.serialize(),
+    options = {
+      scroll: '10m',
+      size: batchSize
+    };
+
+  ndjsonStream.on('data', line => writeStream.write(line));
+
+  await writeData(ndjsonStream, { index, collection });
+  let
+    promises = [],
+    results = await sdk.document.search(index, collection, {}, options);
+
+  do {
+    process.stdout.write(`  ${results.fetched}/${results.total} documents dumped`);
+    process.stdout.write('\r');
+
+    for (const hit of results.hits) {
+      promises.push(writeData(ndjsonStream, {
```

Comment:
Ah, I didn't saw that part when I wrote [this comment](https://github.com/kuzzleio/kuzzle/pull/1319/files#r299345682).

Seeing this, I now think that you should definitely completely remove the `writeData` function and directly invoke `write` here, yielding only if it returns `false`, until the drain event is triggered. That would make writing a dump much, much faster (and simpler)

Reply:
104 sec to 20 sec for 1 000 000 documents, indeed :+1: